### PR TITLE
[data grid] Allow to limit to one filter per column

### LINKED ISSUE #4872: [data grid] Allow to limit to one filter per column
### Duplicates

- [X] I have searched the existing issues

### Latest version

- [X] I have tested the latest version

### Summary 💡

Ideally, I would be able to pass in a prop somewhere that would prevent the filter panel from allowing more than one filter per column.

I could see this prop being on the main DataGridPro component, passed to the filter panel through `componentProps`, or as some kind of field as part of a column definition.

When this setting is enabled and a filter is created for a column this could either remove that column as an option from the column select or perhaps simply disable that column's item within the select.

### Examples 🌈

```JSX
<DataGridPro
    ...
    isOneFilterPerColumn
    // or
    componentsProps={{ filterPanel: { isOneFilterPerColumn: true }}}
    ...
/>
```

Something like that anyway.

### Motivation 🔦

The route I hit to get data for this table only allows one filter per column. An example call might look like: `/some/api/route?titleContains=something&appName=somethingElse`... etc. It can do filtering on multiple columns at once but not multiple filters per column.

I can tailor the column filters to match what is available on the route but I can't prevent the user from creating multiple filters for a column unless I restrict them to only one filter in general which doesn't meet the requirements for the feature I'm implementing.

I have tried manually pruning the filterModel but that runs into issues where you either soft lock the filter panel from creating new filters or end up doing something which would end up just looking like a bug to the user.

### Order ID 💳 (optional)

_No response_ (I'll update when I hear back on what ours is.)

### PR DESCRIPTION
**Pull request by MBilalShafi**
Fixes #4872

Preview: https://deploy-preview-6333--material-ui-x.netlify.app/x/react-data-grid/filtering/#one-filter-per-column

### CODE REVIEW

File: packages/grid/x-data-grid/src/components/panel/filterPanel/GridFilterForm.tsx [hunk 1]
```diff
@@ -10,13 +10,20 @@ import { capitalize, unstable_useId as useId } from '@mui/material/utils';
...
 import { GridFilterItem, GridLinkOperator } from '../../../models/gridFilterItem';
 import { useGridApiContext } from '../../../hooks/utils/useGridApiContext';
 import { useGridRootProps } from '../../../hooks/utils/useGridRootProps';
 import { DataGridProcessedProps } from '../../../models/props/DataGridProps';
 import { getDataGridUtilityClass } from '../../../constants/gridClasses';
-import { GridColDef } from '../../../models/colDef/gridColDef';
+import { GridColDef, GridStateColDef } from '../../../models/colDef/gridColDef';
+
+export interface FilterColumnsArgs {
+  field: string;
```
Comment (cherniavskii) [inline packages/grid/x-data-grid/src/components/panel/filterPanel/GridFilterForm.tsx:23]:
```suggestion
  field: GridColDef['field'];
```
To make it clear what `field` is

File: packages/grid/x-data-grid/src/components/panel/filterPanel/GridFilterForm.tsx [hunk 2]
```diff
@@ -10,13 +10,20 @@ import { capitalize, unstable_useId as useId } from '@mui/material/utils';
...
+  columns: GridStateColDef<any, any, any>[];
```
Comment (cherniavskii) [inline packages/grid/x-data-grid/src/components/panel/filterPanel/GridFilterForm.tsx:24]:
```suggestion
  columns: GridStateColDef[];
```
We can keep the default values

File: packages/grid/x-data-grid/src/components/panel/filterPanel/GridFilterForm.tsx [hunk 3]
```diff
@@ -60,6 +67,12 @@ export interface GridFilterFormProps {
    * @param {GridFilterItem} item The deleted [[GridFilterItem]].
    */
   deleteFilter: (item: GridFilterItem) => void;
+  /**
+   * Allows to filter the columns displayed in the filter form.
+   * @param {FilterColumnsArgs} args The columns of the grid and name of field.
+   * @returns {string[]} The filtered fields array.
+   */
+  filterColumns?: (args: FilterColumnsArgs) => string[];
```
Comment (cherniavskii) [inline packages/grid/x-data-grid/src/components/panel/filterPanel/GridFilterForm.tsx:75]:
```suggestion
   * @returns {GridColDef['field'][]} The filtered fields array.
   */
  filterColumns?: (args: FilterColumnsArgs) => GridColDef['field'][];
```

File: packages/grid/x-data-grid/src/components/panel/filterPanel/GridFilterPanel.tsx [hunk 4]
```diff
@@ -12,13 +12,25 @@ import { useGridRootProps } from '../../../hooks/utils/useGridRootProps';
 import { useGridSelector } from '../../../hooks/utils/useGridSelector';
 import { gridFilterModelSelector } from '../../../hooks/features/filter/gridFilterSelector';
 import { gridFilterableColumnDefinitionsSelector } from '../../../hooks/features/columns/gridColumnsSelector';
+import { GridStateColDef } from '../../../models/colDef/gridColDef';
+
+export interface GetColumnForNewFilterArgs {
+  currentFilters: GridFilterItem[];
+  columns: GridStateColDef<any, any, any>[];
```
Comment (cherniavskii) [inline packages/grid/x-data-grid/src/components/panel/filterPanel/GridFilterPanel.tsx:19]:
```suggestion
  columns: GridStateColDef[];
```

File: packages/grid/x-data-grid/src/components/panel/filterPanel/GridFilterPanel.tsx [hunk 5]
```diff
@@ -12,13 +12,25 @@ import { useGridRootProps } from '../../../hooks/utils/useGridRootProps';
...
   /**
    * The system prop that allows defining system overrides as well as additional CSS styles.
    */
   sx?: SxProps<Theme>;
+  /**
+   * Function that returns the next filter item to be picked as default filter
+   * @param {GetColumnForNewFilterArgs} args Currently configured filters and columns
+   * @returns {string} The field to be used as next filter
+   */
+  getColumnForNewFilter?: (args: GetColumnForNewFilterArgs) => string;
```
Comment (cherniavskii) [inline packages/grid/x-data-grid/src/components/panel/filterPanel/GridFilterPanel.tsx:33]:
```suggestion
   * @returns {GridColDef['field']} The field to be used as next filter
   */
  getColumnForNewFilter?: (args: GetColumnForNewFilterArgs) => GridColDef['field'];
```

File: packages/grid/x-data-grid/src/components/panel/filterPanel/GridFilterPanel.tsx [hunk 6]
```diff
@@ -12,13 +12,25 @@ import { useGridRootProps } from '../../../hooks/utils/useGridRootProps';
...
 export interface GridFilterPanelProps
   extends Pick<GridFilterFormProps, 'linkOperators' | 'columnsSort'> {
   /**
    * The system prop that allows defining system overrides as well as additional CSS styles.
    */
   sx?: SxProps<Theme>;
+  /**
+   * Function that returns the next filter item to be picked as default filter
+   * @param {GetColumnForNewFilterArgs} args Currently configured filters and columns
+   * @returns {string} The field to be used as next filter
```
Comment (cherniavskii) [inline packages/grid/x-data-grid/src/components/panel/filterPanel/GridFilterPanel.tsx:31]:
```suggestion
   * @returns {string} The field to be used for the next filter
```

File: packages/grid/x-data-grid/src/components/panel/filterPanel/GridFilterPanel.tsx [hunk 7]
```diff
@@ -38,6 +50,12 @@ export interface GridFilterPanelProps
   children?: React.ReactNode;
 }
 
+const getGridFilter = (col: GridStateColDef<any, any, any>): GridFilterItem => ({
```
Comment (cherniavskii) [inline packages/grid/x-data-grid/src/components/panel/filterPanel/GridFilterPanel.tsx:53]:
```suggestion
const getGridFilter = (col: GridStateColDef): GridFilterItem => ({
```

**Review by cherniavskii — APPROVED** [review]
Nice addition to filtering :+1:

File: packages/grid/x-data-grid-pro/src/tests/filtering.DataGridPro.test.tsx [hunk 8]
```diff
@@ -69,6 +72,53 @@ describe('<DataGridPro /> - Filter', () => {
     ],
   };
 
+  it('componentsProps `filterColumns` and `getColumnForNewFilter` should allow custom filtering', () => {
```
[the remark this answers is not in the document]
Comment (MBilalShafi) [reply packages/grid/x-data-grid-pro/src/tests/filtering.DataGridPro.test.tsx:75]: Added two more tests to test the two newly added functions, also kept this one as I feel it'll serve as an additional check to make sure the requested usecase is working as expected

**Comment by DanailH** [comment]
An interesting case is: If there is a grid with only 1 column (either the others are hidden or just not present) and this option is enabled what will the filter panel look like in this case? I guess you won't be able to add a new filter altogether (the button is disabled), but I'm not sure how intuitive this is. Currently, the button is enabled but when you click it nothing happens.

You can check a demo here: https://codesandbox.io/s/inspiring-volhard-tnkbos 

Actually even now if you have hidden columns you can still add filters for them. This looks a bit weird but it can be addressed in a different PR.

File: packages/grid/x-data-grid/src/components/panel/filterPanel/GridFilterPanel.tsx [hunk 9]
```diff
@@ -68,40 +87,61 @@ const GridFilterPanel = React.forwardRef<HTMLDivElement, GridFilterPanelProps>(
...
+      return getGridFilter(nextColumnWithOperator);
+    }, [filterModel.items, filterableColumns, getColumnForNewFilter, getDefaultFilter]);
+
     const items = React.useMemo<GridFilterItem[]>(() => {
       if (filterModel.items.length) {
         return filterModel.items;
       }
 
-      const defaultItem = getDefaultItem();
+      const defaultFilter = getDefaultFilter();
```
[the remark this answers is not in the document]
Comment (MBilalShafi) [reply packages/grid/x-data-grid/src/components/panel/filterPanel/GridFilterPanel.tsx:144]: Yeah, the concern is valid, thanks for pointing out.

I tried adding here but it doesn't seem very convincing to call it exactly here as we have to send `defaultFilter` in arguments to `getColumnForNewFilter` on the second filter addition as at that time `filterModel.items` is empty and user has no way to check if a filter is already added. (An exceptional use case)
But I have accommodated `getColumnForNewFilter` in `getDefaultFilter` to address that use case when user has to override the very first field (default one), and I think it works as expected now.

**Comment by nbutenko** [comment]
Hello!
Could u please clarify, is this already in prod?
I'm trying to use the demo code, but it says, that there are no exported values `{ FilterColumnsArgs, GetColumnForNewFilterArgsin }` in x-data-grid-pro

**Comment by MBilalShafi** [comment]
Hey @nbutenko!
It's available in the v6, you can see the documentation [here](https://next.mui.com/x/react-data-grid/filtering/#one-filter-per-column) or playaround with it in this codesandbox [example](https://codesandbox.io/s/sc5p1e?file=/demo.tsx).

MUI-X v6 is currently in beta and will soon be available as a stable release, if you are still using an earlier version, I encourage you to be one of the early adapters to enjoy exciting [new features](https://next.mui.com/x/whats-new/) and a more stable grid experience.

For more information on how to migrate, you can check out our [migration guide](https://next.mui.com/x/migration/migration-data-grid-v5/).

**Comment by nbutenko** [comment]
@MBilalShafi Thank you for sharing!

@cherniavskii That would be great - we're in need for it in our application. We'll keep tracking the progress on this topic!
Appreciate your help!

**Comment by cherniavskii** [comment]
Hi @nbutenko 
I discussed cherry-picking this PR to v5 with the team and we decided not to proceed with it since our release plan for v5 was to backport bug fixes only.

But good news - we're going to release the v6 stable in the next 2-3 weeks and we encourage you to migrate to it once it's stable 🙂

Sorry for the confusion!

**Comment by nbutenko** [comment]
@cherniavskii  Thank you for clarification here! I was about to ask, when I saw the closed PR.
Waiting for the v6! Thank you!

**Pull request merged**